diffkemp/simpll/PatternSet.cpp

```diff
@@ -0,0 +1,420 @@
+                Metadata.GroupEnd = true;
+            } else if (TypeName == "disable-name-comparison") {
+                // disable-name-comparison metadata: string type.
+                Metadata.DisableNameComparison = true;
+            } else if (TypeName == "not-an-input") {
+                // not-an-input metadata: string type.
+                Metadata.NotAnInput = true;
+            } else {
+                DEBUG_WITH_TYPE(DEBUG_SIMPLL,
+                                dbgs() << getDebugIndent()
+                                       << "Invalid metadata type " << TypeName
+                                       << " in node " << *InstMetadata
+                                       << ".\n");
+                return Optional<PatternMetadata>{};
+            }
+            // Shift the operand offset accordingly.
+            OperandIndex += PatternMetadata::MetadataOffsets.lookup(TypeName);
+        } else {
+            return Optional<PatternMetadata>{};
+        }
+    }
+    return {Metadata};
+}
+
+/// Load the given LLVM IR based difference pattern YAML configuration.
+void PatternSet::loadConfig(std::string &ConfigPath) {
+    auto ConfigFile = MemoryBuffer::getFile(ConfigPath);
+    if (std::error_code EC = ConfigFile.getError()) {
+        DEBUG_WITH_TYPE(
+                DEBUG_SIMPLL,
+                dbgs() << getDebugIndent() << "Failed to open difference "
+                       << "pattern configuration " << ConfigPath << ".\n");
+        return;
+    }
+
+    // Parse the configuration file.
+    Input YamlFile(ConfigFile.get()->getBuffer());
+    PatternConfiguration Config;
+    YamlFile >> Config;
+
+    if (YamlFile.error()) {
+        DEBUG_WITH_TYPE(
+                DEBUG_SIMPLL,
+                dbgs() << getDebugIndent() << "Failed to parse difference "
+                       << "pattern configuration " << ConfigPath << ".\n");
+        return;
+    }
+
+    // Load all pattern files included in the configuration.
+    for (auto &PatternFile : Config.PatternFiles) {
+        addPattern(PatternFile);
+    }
+}
+
+/// Add a new LLVM IR difference pattern file.
+void PatternSet::addPattern(std::string &Path) {
+    // Try to load the pattern module.
+    SMDiagnostic err;
+    auto PatternModule = parseIRFile(Path, err, PatternContext);
+    if (!PatternModule) {
+        DEBUG_WITH_TYPE(DEBUG_SIMPLL,
+                        dbgs() << getDebugIndent()
+                               << "Failed to parse difference pattern "
+                               << "module " << Path << ".\n");
+        return;
+    }
+
+    for (auto &Function : PatternModule->getFunctionList()) {
+        // Select only defined functions that start with the left prefix.
+        if (!Function.isDeclaration()
+            && Function.getName().startswith_lower(FullPrefixL)) {
+            StringRef Name = Function.getName().substr(FullPrefixL.size());
+            std::string NameR = FullPrefixR;
+            NameR += Name;
```

Comment:
Then why not:
```suggestion
            std::string Name = Function.getName().substr(FullPrefixL.size());
            std::string NameR = FullPrefixR + Name;
```
You're using `Name.str()` everywhere below anyway.

Reply:
You're right! Did not notice we don't need to use `StringRef` at all here.